### PR TITLE
refactor: allow relative symlinks that points out of the treeartifact

### PR DESCRIPTION
**Pull request by thesayyn**
Fixes https://github.com/bazelbuild/bazel/issues/20891

### CODE REVIEW

**Comment by thesayyn** [comment]
cc @tjgq @comius

**Comment by comius** [comment]
> cc @tjgq @comius

Thanks. I believe @tjgq is more qualified reviewer in this case.

**Review by tjgq — CHANGES_REQUESTED** [review]
LGTM generally, but let's leave some tests in place to catch regressions.

**Comment by tjgq** [comment]
df07d275ce17f14224af0f1a5862755a7db44397 touched some code around this area, so you'll have to rebase your PR.

**Comment by thesayyn** [comment]
All right. I'll get to it soon hopefully.

**Comment by thesayyn** [comment]
@tjgq dangling symlinks isn't seem to be supported anymore. is that expected?  i get tests failing with "is a dangling symbolic link"

**Comment by thesayyn** [comment]
Also would it be possible to cherry pick this into 7.1.0? We'd like to start rolling this out with rules_oci as soon as we can.

File: src/test/java/com/google/devtools/build/lib/skyframe/TreeArtifactBuildTest.java [hunk 1]
```diff
@@ -502,73 +502,6 @@ void run(ActionExecutionContext actionExecutionContext) throws IOException {
...
-    assertThrows(BuildFailedException.class, () -> buildArtifact(out));
-    List<Event> errors = ImmutableList.copyOf(eventCollector);
-    assertThat(errors).hasSize(2);
-    assertThat(errors.get(0).getMessage())
-        .contains(
-            "A TreeArtifact may not contain relative symlinks whose target paths traverse "
-                + "outside of the TreeArtifact");
-    assertThat(errors.get(1).getMessage()).contains("not all outputs were created or valid");
-  }
-
```
[the remark this answers is not in the document]
Comment (thesayyn) [reply src/test/java/com/google/devtools/build/lib/skyframe/TreeArtifactBuildTest.java]: Makes sense. Done. Added the tests back and changed them to assert those cases are still accepted.

**Comment by thesayyn** [comment]
@tjgq Ready for another pass.

**Review by tjgq — APPROVED** [review]

**Comment by tjgq** [comment]
@bazel-io fork 7.1.0

**Pull request closed without merging**